### PR TITLE
[wasm] use sbrk instead of emscripten mmap or malloc when loading bytes into heap

### PR DESCRIPTION
**Pull request by kg**
Emscripten's implementations of `mmap` and `malloc` have a lot of overhead, particularly time spent calling `memset` to zero brand new pages it got by growing the heap. During startup we lose a lot of time to this, because we have to reserve space for things like the sgen card tables and the contents of all our dll files (they live in the heap for metadata decoding).

This PR tries using `sbrk` instead in the hot paths for improved startup, and based on my measurements, it makes most of the time spent in `memset` go away. It's theoretically a breaking change though since someone could have been relying on the ability to `_free` the result of `mono_wasm_load_bytes_into_heap`'d preloaded assets, and our code might be relying on the ability to release pages to the "OS" in wasm?

EDIT: PR updated to not replace all our uses of mmap, and only replace `load_bytes_into_heap` for now - the rest of them will take more work to replace.

### CODE REVIEW

**Comment by kg** [comment]
As suspected it looks like mono_vfree does need to work on wasm, even though there's not really much point in handing pages back (the heap can't ever shrink and there's no wasm equivalent of MADV_DONTNEED).

File: src/mono/browser/runtime/assets.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ export function instantiate_asset(asset: AssetEntry, url: string, bytes: Uint8Ar
         // falls through
         case "heap":
         case "icu":
-            offset = mono_wasm_load_bytes_into_heap(bytes);
+            offset = mono_wasm_load_bytes_into_heap_persistent(bytes);
```
Comment (pavelsavara) [inline src/mono/browser/runtime/assets.ts:40]: What is the sbrk page size or granularity ?
Are there many assets which could be smaller than the size of the sbrk page ?

**Review by pavelsavara — APPROVED** [review]

**Comment by SingleAccretion** [comment]
> `mmap` and `malloc` have a lot of overhead, particularly time spent calling memset to zero brand new pages it got by growing the heap

It is expected that the `mmap` emulation needs to zero pages, but why does `malloc` do this? It looks like a bug. `malloc` doesn't need to return zero-ed memory.

File: src/mono/browser/runtime/assets.ts (hunk 1 above)
Comment (kg) [reply src/mono/browser/runtime/assets.ts:40]: 64k, we could use malloc for small sizes

**Comment by kg** [comment]
> > `mmap` and `malloc` have a lot of overhead, particularly time spent calling memset to zero brand new pages it got by growing the heap
> 
> It is expected that the `mmap` emulation needs to zero pages, but why does `malloc` do this? It sounds like a bug. `malloc` doesn't need to return zeroed memory.

During startup malloc == mmap

**Comment by kg** [comment]
Experimenting with reconfiguring dlmalloc on wasm to use sbrk instead of emscripten mmap/munmap, since they're fake.

**Comment by kg** [comment]
Since we determined that we shouldn't be using dlmalloc at all, removing the dlmalloc change. It did work, though.

**Comment by pavelsavara** [comment]
I can see `--zero-filled-memory` in `wasm-opt` args on some of my CI lane, maybe that makes difference ?

```
Wasm.Console.Bench.Sample -> /__w/1/s/src/mono/sample/wasm/browser-bench/Console/bin/browser-wasm/Wasm.Console.Bench.Sample.dll
  Compiling native assets with /usr/local/emscripten/emsdk/upstream/emscripten/emcc with -Oz. This may take a while ...
  [1/4] corebindings.c -> corebindings.o [took 0.32s]
  [2/4] pinvoke.c -> pinvoke.o [took 0.33s]
  [3/4] driver.c -> driver.o [took 0.34s]
  [4/4] runtime.c -> runtime.o [took 0.36s]
  Linking for initial memory $(WasmInitialHeapSize)=33554432 bytes. Set this msbuild property to change the value.
  Linking with emcc with -O2. This may take a while ...
   "/usr/local/emscripten/emsdk/node/14.18.2_64bit/bin/node" /usr/local/emscripten/emsdk/upstream/emscripten/src/compiler.js /tmp/tmpiu3cpv7n.json --symbols-only
   "/usr/local/emscripten/emsdk/upstream/bin/wasm-ld" -o /__w/1/s/artifacts/obj/mono/Wasm.Console.Bench.Sample/browser.wasm.Release/browser-wasm/wasm/for-build/dotnet.native.wasm /__w/1/s/artifacts/obj/mono/Wasm.Console.Bench.Sample/browser.wasm.Release/browser-wasm/wasm/for-build/pinvoke.o /__w/1/s/artifacts/obj/mono/Wasm.Console.Bench.Sample/browser.wasm.Release/browser-wasm/wasm/for-build/driver.o /__w/1/s/artifacts/obj/mono/Wasm.Console.Bench.Sample/browser.wasm.Release/browser-wasm/wasm/for-build/corebindings.o /__w/1/s/artifacts/obj/mono/Wasm.Console.Bench.Sample/browser.wasm.Release/browser-wasm/wasm/for-build/runtime.o /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.browser-wasm/Release/runtimes/browser-wasm/native/libicudata.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.browser-wasm/Release/runtimes/browser-wasm/native/libicui18n.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.browser-wasm/Release/runtimes/browser-wasm/native/libicuuc.a /__w/1/s/artifacts/bin/microsoft.netcore.ap...
   "/usr/local/emscripten/emsdk/node/14.18.2_64bit/bin/node" /usr/local/emscripten/emsdk/upstream/emscripten/src/compiler.js /tmp/tmp1vo_lkp4.json
   "/usr/local/emscripten/emsdk/upstream/bin/wasm-opt" --post-emscripten -O2 --low-memory-unused --zero-filled-memory --pass-arg=directize-initial-contents-immutable --strip-producers /__w/1/s/artifacts/obj/mono/Wasm.Console.Bench.Sample/browser.wasm.Release/browser-wasm/wasm/for-build/dotnet.native.wasm -o /__w/1/s/artifacts/obj/mono/
...
```

**Comment by pavelsavara** [comment]
I'm bit sad we didn't see this to pass
image

**Comment by kg** [comment]
> I'm bit sad we didn't see this to pass image

It passed a few times before, so I was pretty comfortable not re-running it

**Pull request merged**